### ISSUE TITLE
Add n-dimensional marker-controlled watershed for tensors

### ISSUE THREAD

**Issue by ternaus** [pytorch/pytorch#193591]
## 🚀 Feature request

PyTorch has no marker-controlled watershed operation for n-dimensional tensors.

The immediate use case is volumetric instance segmentation. A model predicts a foreground mask and a boundary score for every voxel. Given integer seeds inside the expected instances, watershed propagates each seed through the foreground while treating high boundary scores as ridges.

The current implementation leaves PyTorch:

```python
labels = skimage.segmentation.watershed(
    boundary_scores.cpu().numpy(),
    markers=markers.cpu().numpy(),
    mask=foreground.cpu().numpy(),
)
labels = torch.from_numpy(labels).to(boundary_scores.device)
```

For a `128 × 256 × 256` volume with uint8 scores, int32 markers, a boolean mask, and int32 output, this moves about 80 MiB between the accelerator and CPU.

## Proposed API

```python
torch.watershed(
    input: Tensor,
    markers: Tensor,
    *,
    mask: Tensor | None = None,
    connectivity: int = 1,
) -> Tensor
```

Proposed semantics:

- `input` contains one real value per spatial position. Lower values are processed first; high values act as ridges.
- `markers` has the same shape and contains zero for unmarked positions and positive instance IDs for seeds.
- `mask` is boolean and restricts where labels may propagate. Output is zero outside the mask.
- `connectivity=1` means four neighbours in 2D and six neighbours in 3D.
- All tensors use the same device. The result has the shape, device, and integer dtype of `markers`.
- All dimensions are spatial; there is no implicit batch or channel dimension.
- The operation is discrete and non-differentiable.

A minimal example:

```text
input:    [ 0,  1,  9,  1,  0]
markers:  [11,  0,  0,  0, 22]
result:   [11, 11, 11, 22, 22]
```

The value `9` acts as a ridge between the two seeds.

For a 3D segmentation pipeline:

```python
instances = torch.watershed(
    contact_logits,  # [D, H, W]
    markers,         # [D, H, W], int32
    mask=foreground, # [D, H, W], bool
    connectivity=1,  # six-connected
)
```

Raw logits can be used directly because watershed depends on their ordering; sigmoid is unnecessary.

## Reference and performance

[`skimage.segmentation.watershed`](https://scikit-image.org/docs/stable/api/skimage.segmentation#skimage.segmentation.watershed) provides the required n-dimensional marker, mask, and connectivity semantics on CPU.

On an Apple M4 Max, scikit-image 0.26 processed a synthetic `128 × 256 × 256` uint8 volume with two markers and an irregular mask in a median of `2.238 s` over five runs.

A PyTorch implementation should be compared with scikit-image across:

- 2D and 3D inputs;
- irregular and disconnected masks;
- different connectivity values;
- plateaus and arbitrary marker IDs;
- uint8, float16, and float32 inputs;
- CPU, CUDA, and MPS.

Accelerator benchmarks must compare the complete native operation with the current device-to-host, scikit-image, and host-to-device route.

## Implementation direction

A CPU implementation can use a stable priority queue, with bucket queues as a possible fast path for integer inputs. CUDA and MPS implementations need device-side queues or frontiers and must not copy through CPU.

A Meta implementation can return a tensor matching `markers`, allowing the operation to remain inside `torch.compile`.

The initial scope does not need automatic marker generation, batching, compact watershed, watershed boundary lines, or autograd.

---

Codex helped draft this issue. I reviewed the API, scope, example, and benchmark because I need this operation in the 3D instance-segmentation inference path described above.
